crates/iota-data-ingestion-core/src/worker_pool.rs

```diff
@@ -92,42 +110,88 @@ impl<W: Worker + 'static> WorkerPool<W> {
+                                if workers[worker_id].send(checkpoint).await.is_err() {
+                                    // The worker channel closing is a sign we need to exit this loop.
                                     break;
                                 }
+                            }
                         }
-                    }
-                    while !checkpoints.is_empty() && !idle.is_empty() {
-                        let checkpoint = checkpoints.pop_front().unwrap();
-                        let worker_id = idle.pop_first().unwrap();
-                        if workers[worker_id].0.send(checkpoint).await.is_err() {
-                            // The worker channel closing is a sign we need to exit this loop.
-                            break;
+                        // Manages the graceful shutdown sequence of the worker pool.
+                        //
+                        // The shutdown process follows these steps:
+                        // 1. Token cancellation triggers workers to:
+                        //    - Complete their current checkpoint processing
+                        //    - Send a final ShutDown message
+                        //    - Exit their processing loop
+                        //
+                        // 2. Main loop continues running to:
+                        //    - Process any remaining Progress messages
+                        //    - Track worker shutdowns via ShutDown messages
+                        //    - Once all workers have reported shutdown, await their join handles
+                        //
+                        // This ensures:
+                        // - No work is interrupted mid-processing
+                        // - All progress messages are processed
+                        // - All resources are properly cleaned up
+                        // - Workers terminate in an orderly fashion
+                        WorkerMsg::ShutDown(worker_id) => {
+                            // Track workers that have initiated shutdown
+                            workers_shutdown_signals.push(worker_id);
+                            // Once all workers have signaled completion, await their handles
+                            // This ensures all workers have finished their final tasks
+                            if workers_shutdown_signals.len() == self.concurrency {
+                                for join_handle in workers_join_handles {
+                                    join_handle.await.expect("worker thread panicked");
+                                }
+                                _ = executor_progress_sender
+                                    .send(WorkerPoolMsg::ShutDown(self.task_name.clone()))
+                                    .await;
+                                tracing::info!("Worker pool `{}` terminated gracefully", self.task_name);
+                                break;
+                            }
```

Comment:
Nit: This can be delegated to a dedicated method that could be called after `tokio_select` in the loop.

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4648/commits/059de2f971efe49b13047a0a3dc70ff0b97a4147#diff-dfef1eeaea282ac0a41b60863611b2244808890824a073bf2264ca6fc775a624R222-R240)